src/ledger/LedgerManagerImpl.cpp

```diff
@@ -854,15 +855,43 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
         try
         {
             xdr::xdr_from_opaque(sv.upgrades[i], lupgrade);
-            Upgrades::applyTo(lupgrade, ledgerDelta.getHeader());
         }
         catch (xdr::xdr_runtime_error)
         {
             CLOG(FATAL, "Ledger") << "Unknown upgrade step at index " << i;
             throw;
         }
+
+        LedgerHeader previousHeader = getCurrentLedgerHeader();
+        try
+        {
+            soci::transaction upgradeScope(getDatabase().getSession());
+            LedgerDelta upgradeDelta(ledgerDelta);
+            Upgrades::applyTo(lupgrade, *this, upgradeDelta);
+            // Note: Index from 1 rather than 0 to match the behavior of
+            // storeTransaction and storeTransactionFee.
+            Upgrades::storeUpgradeHistory(*this, lupgrade,
+                                          upgradeDelta.getChanges(), i + 1);
+            upgradeDelta.commit();
+            upgradeScope.commit();
+        }
+        catch (std::runtime_error& e)
+        {
+            CLOG(ERROR, "Ledger") << "Exception during upgrade: " << e.what();
+            getCurrentLedgerHeader() = previousHeader;
+        }
+        catch (...)
+        {
+            CLOG(ERROR, "Ledger") << "Unknown exception during upgrade";
+            getCurrentLedgerHeader() = previousHeader;
+        }
     }
+    // WARNING: At this point, LedgerManager.getCurrentLedgerHeader() may have
+    // been modified directly as part of the upgrade process. Nothing should be
+    // added between here and the following call to
+    //     ledgerDelta.commit();
 
+    getCurrentLedgerHeader() = initialHeader;
```

Comment:
The issue is that this is functionally equivalent to removing the check completely: if a bug is introduced while processing transactions, it will go unnoticed because of that code.

If you're worried about multiple upgrades doing the wrong thing (which with the current implementation can happen), then we should just
* add a new `LedgerDelta` to wrap all upgrades (and perform the "override" right after the loop, but scoped to upgrades only).
* restore `ledgerManager.getCurrentLedgerHeader()` to its value saved right before performing upgrades

Also, you should add a comment in `Upgrades::applyVersionUpgrade` that
`    ledgerManager.getCurrentLedgerHeader().ledgerVersion = newVersion;`
is basically violating the rule of only using `ledgerDelta.getHeader()` (for future you when you rebase the data layer changes).